Added function to util.py to create and set wedge boundaries for axes

hCraker opened the pull request:
Closes #8 

This PR is an extension of PR #9 which had to be closed due to some issues with my fork that necessitated me deleting it.

Reminder:
Once this PR is merged, an issue in GeoCAT-examples can be resolved (NCAR/GeoCAT-examples#91)

clyne commented:
@hCraker we seem to have lost your earlier logs, but I agree with your comment about simplifying things by making the utility specific to Polar Stereographic and Lambert Conformal projections. If the need arises later, we can always further generalize it.

erogluorhan commented:
@hCraker it is currently a draft (i.e. Work in Progress) PR; please convert it to regular PR when it is actually ready for review.

hCraker commented:
I have renamed the function to something more appropriate `set_map_boundary` since the boundary shape isn't always a true wedge depending the coordinates given and the projection used. I have also tried to find many of the edge cases, but I may not have caught all of them. Please let me know if there is a region of the world or a set of coordinates that produces unexpected behavior. Thanks!

hCraker commented on src/util.py:
```diff
@@ -260,6 +260,95 @@ def xr_add_cyclic_longitudes(da, coord):
+            The axes to which the boundary will be applied.
+
+        projection (:class:'cartopy.crs'):
+            The projection of ax.
```
I think I can actually remove this argument now. While I was writing the function, I used this to deal with various projections differently. Now that we are limiting this to Lambert Conformal and Polar Stereographic, I can treat those the same way and no longer need the if statement that used `projection`.

erogluorhan replied:
Sounds good.

Just a quick and early reminder: Please try to finalize the rest of this function as mature as possible (at least for the current two projections of preference). In other words, try to think of close-to-permanent solutions than temporary apporaches as much as possible. This is because, **geocat-viz** serves as an API to **geocat-examples** and is distributed as an independent package via conda. That being said, a newer version of it is going to be generated when this PR is merged, and **geocat-examples** users will update to that version to use this PR's capabilities. So, we should try to design this utility function to make it as mature as possible.

hCraker replied:
Understood. I was originally trying to make this as general of a function as possible, but I quickly realized that it would be too difficult and unnecessary to support so many projections. I was attempting to do too much at once, the code got messy and, I forgot to remove things. Sorry about that. I'll try to finalize this as much as possible before I re-request your review.

erogluorhan replied:
Totally fine, and we are working on this together. So, we can decide together whenever it is ready for being put into a newer version.

hCraker commented on src/util.py:
```diff
@@ -260,6 +260,95 @@ def xr_add_cyclic_longitudes(da, coord):
+    Utility function to set the boundary of ax to a path that surrounds a
+    given region specified by latitude and longitude coordiantes. This
+    boundary is drawn in the projection coordinates and therefore follows
+    any curves created by the projection. This is currently designed for
```
I have changed it to the following
````
Utility function to set the boundary of ax to a path that surrounds a
given region specified by latitude and longitude coordiantes. This
boundary is drawn in the projection coordinates and therefore follows
any curves created by the projection. As of now, this only works
consistently for the Lambert Conformal Projection and North/South
Polar Stereographic Projections.

erogluorhan replied:
Sounds good. Correct a typo "coordiantes" in "... latitude and longitude coordiantes."

hCraker replied:
I have just pushed this fix

hCraker commented on src/util.py:
```diff
@@ -260,6 +260,95 @@ def xr_add_cyclic_longitudes(da, coord):
+        projection (:class:'cartopy.crs'):
+            The projection of ax.
+
+        lon_range (:class:'tuple'):
```
Good idea. I have made that change. Should I change the class to `(:class:'tuple\list'):` or should I leave it as is?

erogluorhan replied:
You may change it like following as in some other functions in this library:

`:class:`tuple` or :class:`list``

hCraker replied:
I have changed the class specification for `lon_range` and `lat_range`

hCraker commented on src/util.py:
```diff
@@ -260,6 +260,95 @@ def xr_add_cyclic_longitudes(da, coord):
+            Both entries must be between [-90 , 90].
+
+        north_pad (:class:'int'):
+            A constant to be added to the second entry in lat_range. Use this
```
I have added the following note to the documentation to explain this. It should cover the motivation for having `north_pad` `south-pad` `west_pad` and `east_pad` as kwargs.
```
Note: Due to the behavior of cartopy's set_extent() function, the curved 
edges of the boundary may be flattened and cut off. To solve this, use the
kwargs north_pad, south_pad, east_pad, and west_pad. These will modify the
coordinates passed to set_extent(). For the Lambert Conformal and Polar
Stereographic projections, typically only north_pad and south_pad are
needed. If attempting to use this function for other projections
(i.e. Othographic) east_pad and west_pad may be needed.

erogluorhan replied:
Sounds good!

hCraker commented:
> Please provide your responses to my comment above with a trimmed-looking plot and inline comments as well. Then, I can give my second review.

@erogluorhan I have responded to them now.

erogluorhan commented:
> > Please provide your responses to my comment above with a trimmed-looking plot and inline comments as well. Then, I can give my second review.
> 
> @erogluorhan I have responded to them now.

Thanks! What is your response to the following:

Additionally, can you provide the correct calling of this function (i.e. arguments and values) to replace `wedge_boundary` function in `NCL_lcmask_1.py` since when I call this utility function with the same inputs we used for `wedge_boundary`, the resulting plot looks as follows:

![image](https://user-images.githubusercontent.com/32553057/85802042-5957d480-b701-11ea-9c63-fafbecada04b.png)

hCraker commented:
> > > Please provide your responses to my comment above with a trimmed-looking plot and inline comments as well. Then, I can give my second review.
> > 
> > 
> > @erogluorhan I have responded to them now.
> 
> Thanks! What is your response to the following:
> 
> Additionally, can you provide the correct calling of this function (i.e. arguments and values) to replace `wedge_boundary` function in `NCL_lcmask_1.py` since when I call this utility function with the same inputs we used for `wedge_boundary`, the resulting plot looks as follows:
> 
> ![image](https://user-images.githubusercontent.com/32553057/85802042-5957d480-b701-11ea-9c63-fafbecada04b.png)

I have modified this function to work with the arguments given in that example. Do you want me to go ahead and change the lcmask_1 code to use this new function?

erogluorhan commented:
> I have modified this function to work with the arguments given in that example. Do you want me to go ahead and change the lcmask_1 code to use this new function? @erogluorhan

No, thanks! The last time I went and modified `lcmask_1` on my own to use `geocat-viz/util.py/set_map_boundary` was when I made that comment initially, and the plot had some errors as I pasted here. So, I wondered now if there is anything else to be known. You seem to think it should be good for using in `lcmask_1` now, so I will try it on my local and let you know.

hCraker commented:
> > I have modified this function to work with the arguments given in that example. Do you want me to go ahead and change the lcmask_1 code to use this new function? @erogluorhan
> 
> No, thanks! The last time I went and modified `lcmask_1` on my own to use `geocat-viz/util.py/set_map_boundary` was when I made that comment initially, and the plot had some errors as I pasted here. So, I wondered now if there is anything else to be known. You seem to think it should be good for using in `lcmask_1` now, so I will try it on my local and let you know.

@erogluorhan Does the function work for your local version now?

erogluorhan commented:
> @erogluorhan Does the function work for your local version now?

Yes, I now ran it with the exact same parameters as previously done in NCL_lcmask_1:

`gvutil.set_map_boundary(ax, lon_range=[-85, 40], lat_range=[20, 80])`

and it generates 99% the same plot. It only has a very little difference than the plot generated by `wedge_boundary` originally: The very bottom of the plot is a bit trimmed compared to what `wedge_boundary` generated; see below in red ellipse (Though, it is hard to notice from only this plot; you may need to overlap this with what is generated previously):

![image](https://user-images.githubusercontent.com/32553057/86943099-3b2ba480-c103-11ea-9879-241347677efe.png)

This may not be an issue, but I just wonder if this is due to the use of any arguments or those new arguments that I did not explicitly set when calling.

hCraker commented:
> > @erogluorhan Does the function work for your local version now?
> 
> Yes, I now ran it with the exact same parameters as previously done in NCL_lcmask_1:
> 
> `gvutil.set_map_boundary(ax, lon_range=[-85, 40], lat_range=[20, 80])`
> 
> and it generates 99% the same plot. It only has a very little difference than the plot generated by `wedge_boundary` originally: The very bottom of the plot is a bit trimmed compared to what `wedge_boundary` generated; see below in red ellipse (Though, it is hard to notice from only this plot; you may need to overlap this with what is generated previously):
> 
> ![image](https://user-images.githubusercontent.com/32553057/86943099-3b2ba480-c103-11ea-9879-241347677efe.png)
> 
> This may not be an issue, but I just wonder if this is due to the use of any arguments or those new arguments that I did not explicitly set when calling.

Thank you for pointing this out. I am aware of this discrepancy which is why I added the padding kwargs. Set the `south_pad` kwarg equal to 1 and it should work.

erogluorhan commented:
> Thank you for pointing this out. I am aware of this discrepancy which is why I added the padding kwargs. Set the `south_pad` kwarg equal to 1 and it should work.

Thanks for the clarification, I got it!

**Some discussion on this:** 

I am now curious that although this padding (i.e. for south and others) weren't used in the `wedge_boundary`, it worked without any flattening at least for this same plot (Though, I have only skimmed through both `wedge_boundary` and `set_map_boundary` codes for this but not haven't investigated deeply). So, do you have any idea why `wedge_boundary` worked correctly without padding?

Second, I also wonder if this flattening artifact has a deterministic pattern (I mean, if it is due to the use of some "marginal" lat or lon values and so if it can be determined). That way, I am curious if we can handle all these four paddings (i.e. south to east) within the function without needing any input from the user?

hCraker commented:
@erogluorhan I'm not sure why `wedge_boundary` and `set_map_boundary` are behaving differently. When called with the same arguments `ax, lon_range=[-85, 40], lat_range=[20, 80]` they pass the same values into `ax.set_extent()`. In `set_map_boundary` the padding kwargs only change the values for `set_extent` to fix the flattening issue. Additionally, the vertices produced and used by each function to make the boundary are the same with the exception that `set_map_boundary` also generates the vertices for the path along longitude lines.

Regarding a pattern for this, I have yet to find one. There are also so many cases to consider, I'm not sure if we will be able to handle them all.

hCraker commented:
@clairefio @michaelavs Could you two review this when you get a chance? Thanks!

clairefio commented on src/util.py:
```diff
@@ -260,6 +260,109 @@ def xr_add_cyclic_longitudes(da, coord):
+                   [(lon_range[0], lat) for lat in range(lat_range[1], lat_range[0] - 1, -res)]
+        path = mpath.Path(vertices)         
+    elif ((lon_range[0] == 180 or lon_range[0] == -180) and (lon_range[1] == 180 or lon_range[1] == -180)):  
+        verts = [(lon, lat_range[0]) for lon in range(0, 360 + 1, res)]
```
If you wanted to, you could change this variable to "vertices" for consistency.

hCraker commented:
> > Thank you for pointing this out. I am aware of this discrepancy which is why I added the padding kwargs. Set the `south_pad` kwarg equal to 1 and it should work.
> 
> Thanks for the clarification, I got it!
> 
> **Some discussion on this:**
> 
> I am now curious that although this padding (i.e. for south and others) weren't used in the `wedge_boundary`, it worked without any flattening at least for this same plot (Though, I have only skimmed through both `wedge_boundary` and `set_map_boundary` codes for this but not haven't investigated deeply). So, do you have any idea why `wedge_boundary` worked correctly without padding?
> 
> Second, I also wonder if this flattening artifact has a deterministic pattern (I mean, if it is due to the use of some "marginal" lat or lon values and so if it can be determined). That way, I am curious if we can handle all these four paddings (i.e. south to east) within the function without needing any input from the user?

@clyne @erogluorhan I have tried to find a pattern in the cases that require using the padding kwargs, but I have been unable to. I am unsure what we could do to handle these cases and how this PR should progress. Any advice?

erogluorhan commented:
> @clyne @erogluorhan I have tried to find a pattern in the cases that require using the padding kwargs, but I have been unable to. I am unsure what we could do to handle these cases and how this PR should progress. Any advice?

Will work on this next week and let you know. We can progress with approving this as is unless we can find a pattern for the time (The current version works well; I just wanted to see if we can come up with a more generalized version where user is not required to input for padding).
